lib/MusicBrainz/Server/Data/Release.pm

```diff
@@ -1004,31 +1004,34 @@ sub merge
     );
 
     if ($merge_strategy == $MERGE_APPEND) {
-        my %positions = %{ $opts{medium_positions} || {} }
+        my %positions_by_medium = %{ $opts{medium_positions} || {} }
             or confess('Missing medium_positions parameter');
 
+        my %new_positions = map { $_ => 1 } values %positions_by_medium;
         my $update_names = defined $opts{medium_names};
         my %names = %{ $opts{medium_names} || {} };
 
-        my @medium_ids = @{ $self->sql->select_single_column_array(
-            'SELECT id FROM medium WHERE release IN (' . placeholders($new_id, @old_ids) . ')',
+        my @existing_mediums = @{ $self->sql->select_list_of_hashes(
+            'SELECT id, position FROM medium WHERE release IN (' . placeholders($new_id, @old_ids) . ')',
             $new_id, @old_ids
         ) };
 
         confess('medium_positions does not account for all mediums in all releases')
-            unless all { exists $positions{$_} } @medium_ids;
+            unless all {
+                exists $positions_by_medium{$_->{id}} || !exists $new_positions{$_->{position}}
```

Comment:
Here's the main change, since this is where the exception is thrown.

The `exists $positions_by_medium{$_->{id}}` by itself is the same as the previous test. But we also now check if the medium we're looking has a position that doesn't conflict with any the edit wants to take.